### ISSUE TITLE
Memory Leak

### ISSUE THREAD

**Issue by mudtoe** [nicolargo/glances#1089]
Recent versions of glances seem to suffer from a memory leak.  Leaving it running for a couple of days will show, on the glances screen itself, glances using more and more memory.

**Comment by nicolargo** [comment]
By default, Glances manages an history of the stats. By default this history is limited to 1 day thanks to the Glances configuration file (see: https://github.com/nicolargo/glances/blob/master/conf/glances.conf#L8). So during one day, the memory will grow up and this is not a memory leak.

Nevertheless, if the memory continue to increase after one day it's a bug...

Can you confirm this last point ?

**Comment by giampaolo** [comment]
What platform is this on? The most probable candidates for a memory leak are third party deps using a C extension, amongst which psutil. psutil has a suite for checking memory leaks which you may wanna try (GIT clone psutil repo then run `make setup-dev-env && make test-memleaks`)

**Comment by mudtoe** [comment]
Sorry for taking so long to respond.  I'm using Arch Linux, and it's happened on several kernel versions.  I'll have to check to see if it grows beyond 24 hours, but it's using about 100mb of memory after a couple of days.  Would stats history ever use that much?   I'll make some screen shots and post the results.   I'm running glances 2.10 with version 5.2.2 of PSutil.

**Comment by mudtoe** [comment]
OK, I've completed my tests on glances.  nicolargo, you are correct in that glances only increases memory usage for 24 hours, and then it stabilizes.  That's the good news.  The bad news is that the increase in memory usage is "reserved" memory, which I believe means it can't be swapped.   In my test glances starts out with about 34mb of reserved memory, and after a day it's about 180mb of reserved memory.  That's a lot for a system or VM without a lot of memory assigned to it.    I'm attaching three screenshots of glances, one shortly after it was started, the second 24 hours later, and the third about 24 hours after that.   You can see how the memory usage stabilized after 24 hours, and you can also see the big increase in reserved memory and the corresponding decrease in free system memory.  
 
I'm going to look into decreasing the amount of history data kept, but it would be really nice if the memory being used for stats was not marked reserved, as that would pretty much make this a non-issue, even on a low memory PC or VM.   
 
If you have any questions, let me know.
 

![glances1](https://user-images.githubusercontent.com/28747682/27267364-d51ce57a-5475-11e7-97bf-3366e2c8ec7c.jpg)
![glances2](https://user-images.githubusercontent.com/28747682/27267366-d51e8196-5475-11e7-87f7-a11852ecbb23.jpg)
![glances3](https://user-images.githubusercontent.com/28747682/27267365-d51e1d96-5475-11e7-9150-31d83b566fce.jpg)

**Comment by Wilkolicious** [comment]
This is glances (Glances v2.1.1 with psutil v5.0.1) running since about 5pm yesterday on a production server:
![screen shot 2017-06-21 at 16 54 06](https://user-images.githubusercontent.com/16191979/27394008-ab367fee-56a2-11e7-9391-004224c4864d.png)
with `history_size=400` in `/etc/glances/glances.conf`

**Comment by Wilkolicious** [comment]
Hmm, found glances stopped in my tmux session with "Killed", so must've died last night...

**Comment by Wilkolicious** [comment]
Glances v2.10 with psutil v5.2.2 with `history_size=400` (ran from `glances -w -C /etc/glances/glances.conf`) running for the past 6h18m:
![screen shot 2017-06-22 at 15 25 07](https://user-images.githubusercontent.com/16191979/27439141-170e46f4-575f-11e7-93fe-e1b98ad402f6.png)

You have a memory leak laddie.

**Comment by nicolargo** [comment]
@mudtoe You can use the history_size in the configuration file to reduce the memory used by Glances.

```
[global]
# History size (maximum number of values)
# Default is 28800: 1 day with 1 point every 3 seconds (default refresh time)
history_size=28800
```

Note: the history_size value *do not* set the maximum memory size but the number of values stored in the memory -used by the graph output plugin) ?

**Comment by mudtoe** [comment]
Yes, from your previous post I got that.  My only remaining point was wondering why keeping stats would require reserved memory that couldn't be swapped.  If the memory glances was using for that wasn't non-swappable, then there would be no issue on low memory systems, as it could be swapped as needed.



I'm not a Linux programmer, but I was for many years an IBM mainframe developer (I worked in assembler) and in that world there were ways (if you were a privileged program) to tell the OS via  an INT call what kind of memory you wanted, including whether or not the memory was eligible to be paged out (the equivalent of being swapped in Linux lingo).  The default was always pageable memory (and the only memory non-privileged programs could request), unless you did something specific on the parameters to the INT call to indicate otherwise.   I guess I was just asking if there was something set in how glances was acquiring memory for these stats that causes it to be reserved and ineligible to be swapped, which could be changed.


Sent from Outlook<http://aka.ms/weboutlook>


________________________________
